tests/tape/test_tape_measure.py

```diff
@@ -501,3 +512,360 @@ def func():
+        def func():
+            qml.Hadamard(0)
+            return density_matrix(0)
+
+        func()
+        obs = func.qtape.observables
+        assert len(obs) == 1
+        assert obs[0].return_type is State
+
+    def test_return_type_is_state_mixed(self):
+        """Test that the return type of the observable is State"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(0)
+            return density_matrix(0)
+
+        func()
+        obs = func.qtape.observables
+        assert len(obs) == 1
+        assert obs[0].return_type is State
+
+    def test_correct_density_matrix_product_state_first_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(0)
+
+        density_first = func()
+
+        assert np.allclose(
+            np.array([[0.0 + 0.0j, 0.0 + 0.0j], [0.0 + 0.0j, 1.0 + 0.0j]]), density_first
+        )
+
+    def test_correct_density_matrix_product_state_first_mixed(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(0)
+
+        density_first = func()
+
+        assert np.allclose(
+            np.array([[0.0 + 0.0j, 0.0 + 0.0j], [0.0 + 0.0j, 1.0 + 0.0j]]), density_first
+        )
+
+    def test_correct_density_matrix_product_state_second_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(1)
+
+        density_second = func()
+        assert np.allclose(
+            np.array([[0.5 + 0.0j, 0.5 + 0.0j], [0.5 + 0.0j, 0.5 + 0.0j]]), density_second
+        )
+
+    def test_correct_density_matrix_product_state_second_mixed(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(1)
+
+        density_second = func()
+        assert np.allclose(
+            np.array([[0.5 + 0.0j, 0.5 + 0.0j], [0.5 + 0.0j, 0.5 + 0.0j]]), density_second
+        )
+
+    def test_correct_density_matrix_all_wires_qubit(self):
+        """Test that the correct density matrix  is returned when
+        all wires are given"""
```

Comment:
A nice test to ensure that the partial trace logic is skipped is to use the `mocker` fixture. In particular, [`mocker.spy`](https://github.com/pytest-dev/pytest-mock/#spy) allows you to 'spy' on a function/method call, and assert whether it has been called or not.